sdks/java/core/src/main/java/org/apache/beam/sdk/options/ProxyInvocationHandler.java

```diff
@@ -210,6 +268,125 @@ public int hashCode() {
   }
 
   /**
+   * Populate display data. See {@link HasDisplayData#populateDisplayData}. All explicitly set
+   * pipeline options will be added as display data.
+   */
+  private void populateDisplayData(DisplayData.Builder builder) {
+    Set<PipelineOptionSpec> optionSpecs = PipelineOptionsReflector.getOptionSpecs(knownInterfaces);
+    Multimap<String, PipelineOptionSpec> optionsMap = buildOptionNameToSpecMap(optionSpecs);
+
+    for (Map.Entry<String, BoundValue> option : options.entrySet()) {
+      BoundValue boundValue = option.getValue();
+      if (boundValue.isDefault()) {
+        continue;
+      }
+
+      Object value = boundValue.getValue() == null ? "" : boundValue.getValue();
+      DisplayData.Type type = DisplayData.inferType(value);
+      HashSet<PipelineOptionSpec> specs = new HashSet<>(optionsMap.get(option.getKey()));
+
+      for (PipelineOptionSpec optionSpec : specs) {
+        Class<?> pipelineInterface = optionSpec.getDefiningInterface();
+        if (type != null) {
+          builder.add(option.getKey(), type, value)
+              .withNamespace(pipelineInterface);
+        } else {
+          builder.add(option.getKey(), value.toString())
+              .withNamespace(pipelineInterface);
+        }
+      }
+    }
+
+    for (Map.Entry<String, JsonNode> jsonOption : jsonOptions.entrySet()) {
+      if (options.containsKey(jsonOption.getKey())) {
+        // Option overwritten since deserialization; don't re-write
+        continue;
+        // TODO: Is it worth removing the JSON option for consistency?
+      }
+      HashSet<PipelineOptionSpec> specs = new HashSet<>(optionsMap.get(jsonOption.getKey()));
+      if (specs.isEmpty()) {
+        builder.add(jsonOption.getKey(), jsonOption.getValue().toString())
+          .withNamespace(UnknownPipelineOptions.class);
+      } else {
+        for (PipelineOptionSpec spec : specs) {
+          Object value = getValueFromJson(jsonOption.getKey(), spec.getGetterMethod());
+          DisplayData.Type type = DisplayData.inferType(value);
+          if (type != null) {
+            builder.add(jsonOption.getKey(), type, value)
+                .withNamespace(spec.getDefiningInterface());
+          } else {
+            builder.add(jsonOption.getKey(), value.toString())
+                .withNamespace(spec.getDefiningInterface());
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Marker interface use when the original {@link PipelineOptions} interface is not avaialble at
+   * runtime. This can occur when {@link PipelineOptions} are deserialized from JSON or specified
+   * on the command line.
+   *
+   * To ensure {@link PipelineOptions} type information is available at runtime, register known
```

Comment:
Don't you really just want to make a sure a call has been made to PipelineOptionsFactory (like getRegisteredOptions)?

This will ensure the @AutoService registration for pipeline options will have happened then.